File: src/network/request/RequestManager.php
<?php

namespace pocketcloud\cloud\network\request;

use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\Network;
use pocketcloud\cloud\network\packet\RequestClientPacket;
use pocketcloud\cloud\network\packet\RequestPacketFailureReason;
use pocketcloud\cloud\network\packet\ResponseClientPacket;
use pocketcloud\cloud\util\misc\Tickable;
use pocketcloud\cloud\util\trait\SingletonTrait;

final class RequestManager implements Tickable {
    use SingletonTrait;

    /** @var array<string, RequestClientPacket> */
    private array $requests = [];

    public function __construct() {
        self::setInstance($this);
    }

    /**
     * @internal
     * @see RequestClientPacket
     */
    public function send(RequestClientPacket $packet, ServerClient $client): RequestClientPacket|false {
        $packet->prepare();
        if (!Network::getInstance()->sendPacket($packet, $client)) return false;
        $this->requests[$packet->getRequestId()] = $packet;
        return $packet;
    }

    public function remove(RequestClientPacket|string $request): void {
        $requestId = $request instanceof RequestClientPacket ? $request->getRequestId() : $request;
        unset($this->requests[$requestId]);
    }

    public function resolve(ResponseClientPacket $packet): void {
        if (isset($this->requests[$packet->getRequestId()])) {
            $RequestClientPacket = $this->requests[$packet->getRequestId()];
            if ($RequestClientPacket instanceof RequestClientPacket) {
                $RequestClientPacket->invokeClosures(false, $packet);
            }
        }
    }

    public function reject(RequestClientPacket $packet): void {
        if (isset($this->requests[$packet->getRequestId()])) {
            $packet->invokeClosures(true, null, RequestPacketFailureReason::REQUEST_TIMEOUT);
        }
    }

    public function tick(int $currentTick): void {
        foreach ($this->requests as $request) {
            if (($request->getSentTimestamp() + 10) < microtime(true)) {
                RequestManager::getInstance()->reject($request);
                RequestManager::getInstance()->remove($request);
            }
        }
    }

    public function get(string $requestId): ?RequestClientPacket {
        return $this->requests[$requestId] ?? null;
    }

    public function getAll(): array {
        return $this->requests;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/network/request/RequestManager.java
package de.pocketcloud.cloud.network.request;

import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.cloud.network.packet.RequestClientPacket;
import de.pocketcloud.cloud.network.packet.RequestPacketFailureReason;
import de.pocketcloud.cloud.network.packet.ResponseClientPacket;
import de.pocketcloud.common.lifecycle.Tickable;
import lombok.Getter;
import lombok.experimental.Accessors;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Tracks pending {@link RequestClientPacket}s sent from the cloud to sub-servers.
 * Resolves them when the matching {@link ResponseClientPacket} arrives, and rejects
 * them when the 10-second timeout expires.
 */
public final class RequestManager implements Tickable {

    private final Map<String, RequestClientPacket> requests = new ConcurrentHashMap<>();

    public RequestClientPacket send(RequestClientPacket packet, ServerClient client) {
        if (requests.containsKey(packet.getRequestId())) return requests.get(packet.getRequestId());
        requests.put(packet.getRequestId(), packet);
        client.sendPacket(packet).exceptionally(e -> {
            reject(packet, e);
            remove(packet);
            return null;
        });

        return packet;
    }

    public void remove(RequestClientPacket packet) {
        remove(packet.getRequestId());
    }

    public void remove(String requestId) {
        requests.remove(requestId);
    }

    public void resolve(ResponseClientPacket packet) {
        RequestClientPacket request = requests.get(packet.getRequestId());
        if (request != null) {
            request.invokeClosures(false, packet, null, null);
            remove(request);
        }
    }

    public void reject(RequestClientPacket packet) {
        packet.invokeClosures(true, null, RequestPacketFailureReason.REQUEST_TIMEOUT, null);
    }

    public void reject(RequestClientPacket packet, Throwable e) {
        packet.invokeClosures(true, null, RequestPacketFailureReason.EXCEPTION, e);
    }

    @Override
    public void tick(long currentTick) {
        long now = System.currentTimeMillis();
        requests.entrySet().removeIf(entry -> {
            RequestClientPacket request = entry.getValue();
            Long sent = request.getSentTimestamp();
            if (sent != null && (now - sent) > 10_000) {
                reject(request);
                return true;
            }

            return false;
        });
    }

    public RequestClientPacket get(String requestId) {
        return requests.get(requestId);
    }

    public Map<String, RequestClientPacket> getAll() {
        return new HashMap<>(requests);
    }
}